### ISSUE TITLE
Compare performance of various mAP metrics

### ISSUE THREAD

**Issue by potipot** [airctic/icevision#654]
## 🚀 Feature
Following the discussion from ice-dev it was suggested to compare different mAP metrics in terms of their performance.
So far what we have available:
 - "default" COCOMetric
 - Frederik's mAP implementation
 - rafaelpadilla's project https://github.com/rafaelpadilla/review_object_detection_metrics 
 - *potentially mAP implemented by folks at pytorch-lightning* https://github.com/PyTorchLightning/pytorch-lightning/pull/4564

**Comment by rsomani95** [comment]
Quick observation: @fstroth and rafaelpadilla's project both use NumPy whereas the lightning implementation uses PyTorch. Curious how much of a speedup that would offer

**Comment by lgvaz** [comment]
> use NumPy whereas the lightning implementation uses PyTorch.
Numpy would be better for us because eventually pytorch should become an optional dependency (we want to start offering tensorflow support). But I'm also curious to see the difference in speed.

**Comment by potipot** [comment]
Hmm, is pytorch faster than numpy when running on CPU?

**Comment by lgvaz** [comment]
> Hmm, is pytorch faster than numpy when running on CPU?

It might be in some cases, but **how** much faster?

**Comment by rsomani95** [comment]
> Numpy would be better for us because eventually pytorch should become an optional dependency (we want to start offering tensorflow support)
Makes sense. Ambitious!

Not to crowd this space, but here's another NumPy implementation, from the author of Albumentations:
https://github.com/ternaus/iglovikov_helper_functions/blob/master/iglovikov_helper_functions/metrics/map.py

**Comment by fstroth** [comment]
If we want to go for maximum performance we should look into writing a version using numba.

**Comment by potipot** [comment]
![image](https://user-images.githubusercontent.com/52150545/109304360-6f8fd480-783c-11eb-9da8-e3f490de85a4.png)
Best way would be to see where we stand at, following the prof Knuth advice :D

**Comment by FraPochetti** [comment]
Not needed anymore.

**Comment by potipot** [comment]
@FraPochetti I'd keep this one in the backlog. 

This is exactly the reason why I got stuck working on custom metric implementation - results of mAP calculation for the same data all mutually differ on direct pycoctools, icevision+COCOMetric, and rafaelpadilla library.
